### ISSUE TITLE
Add axe localisations to NPM

### ISSUE THREAD

**Issue by WilcoFiers** [dequelabs/axe-core#441]
Today, if you want to get the Japanese localisation, you'll have to clone axe-core, install it, and build that file yourself. That's not particularly useful. Ideally, all localisations get included into the published version of axe-core, so other projects like axe-cli and axe-webdriverjs can include the correct language.

**Comment by kurosawa-takeshi** [comment]
Though there are number of possible ways to do, I think there are three types of approaches as follows. I think that Approach B is the most realistic.

# Approach A

- Adding axe.ja.min.js, axe.nl.js and so on to the npm package.
    
    Use `grunt build --all-lang` instead of `grunt build` when creating the npm package.

## Pros.

- Additional works are not needed.
- It will be suited for scenarios using script element.
    
    Ex. `<script src='node_modules/axe-core/axe.ja.min.js' ></script>`.

## Cons.

- It will not be suited for add-ons and WebDriver.
    
    It is not guaranteed that there is `axe.${chrome.i18n.getUILanguage()}.min.js`.
    
    Negotiating locales which browsers and servers do is needed. However, axe doesn’t have this kind of APIs at this moment.


# Approach B

- Extending APIs and keeping backward compatibility of axe APIs

## Pros.

- It will be suited for more scenarios including add-ons and WebDriver

## Cons.

- Additional works are needed

## Suggested API changes

I suggest that adding `locale` to `axe.run`'s options to specify locale of the results.

```javascript
axe.run({ options: { locale: 'ja' } }).then(({ locale /* new */, violations }) => {
    // Japanese is supported so that Japanese is used
    locale === 'ja'; // => true

    // volations[0].description => 'イメージマップの<area>要素に...' 
    // Note: 'イメージマップの<area>要素に...'  is Japanese translation of 'Checks the <area> elements of image...'
});

axe.run({ options: { locale: 'fr' } }).then(({ locale /# new #/, violations }) => {
    // French is not supported so that default locale (English) is used
    locale === 'en'; // => true

    // violations[0].description => 'Checks the <area> elements of image...' 
});


// axe.run could be ...
axe.run = function(...) {
    // ...

    // Negotiating locales
    // ref. https://github.com/florrain/locale/
    const locale = chooseBestLocale(options.locale);

    if (axe.getCurrentLocale() !== locale) {
        // Replace all messages to localized messages before executing axe._runRules
    }

    axe._runRules(...);

    return p;
}
```

Misc. APIs which handle locales are also needed.

```javascript
// New APIs for locales
axe.getLocales() === ['en', 'ja', 'nl']); // => true

axe.getDefaultLocale() === 'en'; // => true

axe.getCurrentLocale() === 'en'; // => true
```

I suggest that add `locale` to `axe.configure` to make it configure current locale.

```javascript
// Demonstrating how locales affect existing APIs

axe.getCurrentLocale() === axe.getDefaultLocale(); // => true

// Default locale’s messages should be used by default.
let rules = axe.getRules();
// rules[0] => { ruleId: 'area-alt', description: 'Checks the <area> elements of image...' },

axe.run().then(({ violations }) => {
    // violations[0].description => 'Checks the <area> elements of image...' });
});


// New option for axe.configure
axe.configure({ locale: 'ja' });

axe.getCurrentLocale() === 'ja'; // => true

// Specified locale’s messages are used.
rules = axe.getRules();
// rules[0] => { ruleId: 'area-alt', description: 'イメージマップの<area>要素に...' },

axe.run().then(({ violations }) => {
    // violations[0].description => 'イメージマップの<area>要素に...' });
});

axe.reset();

axe.getCurrentLocale() === axe.getDefaultLocale(); // => true

// Default locale’s messages are used.
rules = axe.getRules();
// rules[0] => { ruleId: 'area-alt', description: 'Checks the <area> elements of image...' },

axe.run().then(({ violations }) => {
    // violations[0].description => 'Checks the <area> elements of image...' });
});

axe.run({ options: { locale: 'ja' } }).then(({ violations }) => {
    // violations[0].description => 'イメージマップの<area>要素に...' });
});

// I think that axe.run should not affect current locale
axe.getCurrentLocale() === axe.getDefaultLocale(); // => true
rules = axe.getRules();
// rules[0] => { ruleId: 'area-alt', description: 'Checks the <area> elements of image...' },
```

# Approach C

- Extending APIs with breaking changes

If we can introduce incompatible API changes, we could do more drastic things like

```javascript
axe.run().then((rawResults) => {
    // `rawResults` is something similar to objects which `reporter` processes
    let resultsForEnglishUsers = axe.translateResult(rawResults);
    let resultsForJapaneseUsers = axe.translateResult(rawResults, 'ja');
});
```

## Pros.

- Great flexibility

## Cons.

- Additional works are needed
- Users need to change their codes

**Comment by WilcoFiers** [comment]
@kurosawa-takeshi Thanks for that analyses. That's very helpful. I'm leaning towards option A. It's the simplest solution by far, and you can actually already use it in axe-webdriverjs and axe-cli as well. You can specify the axe source you want to use, so if you point it to axe.ja.js it should run with that. One addition I'd like to have for it is to modify the current locale file format so that you can run it in axe.configure(). This shouldn't be too difficult, and would mean you can either run `axe.ja.js`, or you can run `axe.js`, pull in `locale/ja.json`, and pass it to `axe.configure()`. What do you think?

**Comment by WilcoFiers** [comment]
Oh! Additionally, one thing to consider is that currently the locale files have a template format which is compiled during the built step. We don't have that parser in the axe source, so we can't do it on runtime. That's the reason that currently we can't currently change languages in runtime.

**Comment by nschonni** [comment]
An approach like B could look at a library like https://github.com/DanielBaulig/node-gettext (never used it) rather than coming up with something custom

**Comment by WilcoFiers** [comment]
Closing the issue. We've got a solution that should work for the time being. There are now two ways to get axe-core results localised. Either by building a localised version of axe-core, or by passing the locale data to `axe.configure({ locale })`.

If there's ever a solid use case for it we could create a result translation utility, but it probably wouldn't be part of axe-core itself. It makes more sense to have that be its own tool.
